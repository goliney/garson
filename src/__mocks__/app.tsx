import React from 'react';
import { render } from 'ink-testing-library';
import { Text } from 'ink';

const initialView = <Text>Initializing tests...</Text>;

export const app = render(initialView);

app.unmount = jest.fn().mockImplementation(() => app.rerender(initialView));
