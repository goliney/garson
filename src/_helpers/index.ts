// Components
export { Question } from './components/question';
export { Line } from './components/line';

// Hooks
export { useKeyHandler } from './use-key-handler';
export { useEnterKeyHandler } from './use-enter-key-handler';
