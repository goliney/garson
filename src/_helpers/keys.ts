export const ARROW_UP = '\u001B[A';
export const ARROW_DOWN = '\u001B[B';
export const ARROW_LEFT = '\u001B[D';
export const ARROW_RIGHT = '\u001B[C';
export const ENTER = '\r';
export const CTRL_C = '\x03';
export const BACKSPACE = '\x08';
export const DELETE = '\x7F';
export const SPACE = ' ';
