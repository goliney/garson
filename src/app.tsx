import { Box, render } from 'ink';
import React from 'react';

export const app = render(<Box>Initializing...</Box>, {
  debug: false,
});
