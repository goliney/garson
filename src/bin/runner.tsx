import React from 'react';
import { Box, render } from 'ink';
import { GarsonConfig } from '../types';

const { rerender, unmount } = render(<Box>Initializing...</Box>, {
  debug: false,
});

export function runner({ results, prompts, action }: GarsonConfig) {
  const [{ prompt, key }, ...rest] = prompts;
  rerender(
    prompt(result => {
      rerender(<React.Fragment />); // clear the previous prompt from the screen
      const newResults = { ...results, [key]: result };
      if (rest.length === 0) {
        // it was the last prompt in a chain
        const actionResponse = action(newResults);
        if (actionResponse) {
          runner(actionResponse);
        } else {
          unmount();
        }
      } else {
        // there is another prompt
        runner({ results: newResults, prompts: rest, action });
      }
    }),
  );
}
