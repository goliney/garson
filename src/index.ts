import { KeyType, PromptCallback, Results, ActionCallback, GarsonHandlerParams } from './types';

export function garson(results?: Results) {
  return garsonHandler({ results });
}

function garsonHandler({ results = {}, prompts = [] }: GarsonHandlerParams) {
  return {
    prompt: (key: KeyType, prompt: PromptCallback) =>
      garsonHandler({ results, prompts: [...prompts, { key, prompt }] }),
    action: (action: ActionCallback) => ({
      results,
      prompts,
      action,
    }),
  };
}
