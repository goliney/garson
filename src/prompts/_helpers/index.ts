// Components
export { Question } from './components/question';

// Misc
export { useKeyHandler } from './use-key-handler';
