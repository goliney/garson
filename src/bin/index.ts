import { runner } from './runner';

const cwd = process.cwd();
let garsonConfig;

try {
  // eslint-disable-next-line import/no-dynamic-require, global-require
  garsonConfig = require(`${cwd}/garson.config.js`);
} catch (e) {
  console.error('Failed to load config', e);
}

if (garsonConfig) {
  runner(garsonConfig);
}
