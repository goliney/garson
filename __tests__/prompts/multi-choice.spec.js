import { garson, prompts } from '../../src';
import { runner } from '../../src/bin/runner';
import { app } from '../../src/app';
import { ENTER, SPACE, ARROW_DOWN, ARROW_UP } from '../../src/_helpers/keys';
import { write, stripColorsFromLastFrame } from '../helpers';

jest.mock('../../src/app');

const actionSpy = jest.fn();

const config = garson()
  .prompt(
    'wcOptions',
    prompts.multiChoices({
      message: 'What do you want to count in garson.config.js file?',
      items: [
        { label: 'Lines', value: 'l', isSelected: true },
        { label: 'Words', value: 'w', isSelected: true },
        { label: 'Characters', value: 'm' },
        { label: 'Everything', value: 'lwm', isSelected: false },
      ],
      // optional handler
      onChangeMiddleware(newItems, oldItems, allItems) {
        const isEverything = item => item.value === 'lwm';
        const newItemsHaveEverythingSelected = newItems.some(isEverything);
        const oldItemsHaveEverythingSelected = oldItems.some(isEverything);

        if (newItemsHaveEverythingSelected && !oldItemsHaveEverythingSelected) {
          // if "Everything" just got selected, deselect everything else
          return allItems.filter(isEverything);
        }

        if (oldItemsHaveEverythingSelected) {
          // if "Everything" was selected, deselect it
          return newItems.filter(item => !isEverything(item));
        }

        return newItems;
      },
    })
  )
  .action(actionSpy);

describe('Multi Choice', () => {
  beforeEach(async () => {
    actionSpy.mockClear();
    runner(config);
    await new Promise(resolve => setTimeout(resolve));
  });

  afterEach(() => {
    app.unmount();
  });

  test('Initial render', () => {
    expect(stripColorsFromLastFrame()).toContain('▇ ◉ Lines');
    expect(app.lastFrame()).toMatchSnapshot();
  });

  test('Arrow controls', async () => {
    await write(ARROW_DOWN);
    expect(stripColorsFromLastFrame()).toContain('▇ ◉ Words');
    expect(app.lastFrame()).toMatchSnapshot();

    await write(ARROW_DOWN);
    expect(stripColorsFromLastFrame()).toContain('▇ ◯ Characters');
    expect(app.lastFrame()).toMatchSnapshot();

    await write(ARROW_DOWN);
    expect(stripColorsFromLastFrame()).toContain('▇ ◯ Everything');
    expect(app.lastFrame()).toMatchSnapshot();

    await write(ARROW_DOWN);
    expect(stripColorsFromLastFrame()).toContain('▇ ◯ Everything');
    expect(app.lastFrame()).toMatchSnapshot();

    await write(ARROW_UP);
    expect(stripColorsFromLastFrame()).toContain('▇ ◯ Characters');
    expect(app.lastFrame()).toMatchSnapshot();

    await write(ARROW_UP);
    expect(stripColorsFromLastFrame()).toContain('▇ ◉ Words');
    expect(app.lastFrame()).toMatchSnapshot();

    await write(ARROW_UP);
    expect(stripColorsFromLastFrame()).toContain('▇ ◉ Lines');
    expect(app.lastFrame()).toMatchSnapshot();

    await write(ARROW_UP);
    expect(stripColorsFromLastFrame()).toContain('▇ ◉ Lines');
    expect(app.lastFrame()).toMatchSnapshot();
  });

  test('Toggle selection', async () => {
    await write(SPACE);
    expect(app.lastFrame()).toMatchSnapshot();

    await write(ARROW_DOWN);
    await write(ARROW_DOWN);
    await write(SPACE);

    expect(app.lastFrame()).toMatchSnapshot();

    await write(ENTER);
    expect(actionSpy).toHaveBeenCalledWith({ wcOptions: ['w', 'm'] });
  });

  test('onChangeMiddleware', async () => {
    await write(ARROW_DOWN);
    await write(ARROW_DOWN);
    await write(ARROW_DOWN);

    await write(SPACE); // toggle 'Everything'
    expect(stripColorsFromLastFrame()).toContain('◯ Lines');
    expect(stripColorsFromLastFrame()).toContain('◯ Words');
    expect(stripColorsFromLastFrame()).toContain('◯ Characters');
    expect(stripColorsFromLastFrame()).toContain('◉ Everything');
    expect(app.lastFrame()).toMatchSnapshot();

    await write(ARROW_UP);
    await write(SPACE); // toggle something besides of 'Everything'
    expect(stripColorsFromLastFrame()).toContain('◯ Lines');
    expect(stripColorsFromLastFrame()).toContain('◯ Words');
    expect(stripColorsFromLastFrame()).toContain('◉ Characters');
    expect(stripColorsFromLastFrame()).toContain('◯ Everything');
    expect(app.lastFrame()).toMatchSnapshot();
  });
});
