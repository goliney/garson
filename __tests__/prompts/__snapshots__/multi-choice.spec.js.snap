// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multi Choice Arrow controls 1`] = `
"? What do you want to count in garson.config.js file?
  ◉ Lines
▇ ◉ Words
  ◯ Characters
  ◯ Everything"
`;

exports[`Multi Choice Arrow controls 2`] = `
"? What do you want to count in garson.config.js file?
  ◉ Lines
  ◉ Words
▇ ◯ Characters
  ◯ Everything"
`;

exports[`Multi Choice Arrow controls 3`] = `
"? What do you want to count in garson.config.js file?
  ◉ Lines
  ◉ Words
  ◯ Characters
▇ ◯ Everything"
`;

exports[`Multi Choice Arrow controls 4`] = `
"? What do you want to count in garson.config.js file?
  ◉ Lines
  ◉ Words
  ◯ Characters
▇ ◯ Everything"
`;

exports[`Multi Choice Arrow controls 5`] = `
"? What do you want to count in garson.config.js file?
  ◉ Lines
  ◉ Words
▇ ◯ Characters
  ◯ Everything"
`;

exports[`Multi Choice Arrow controls 6`] = `
"? What do you want to count in garson.config.js file?
  ◉ Lines
▇ ◉ Words
  ◯ Characters
  ◯ Everything"
`;

exports[`Multi Choice Arrow controls 7`] = `
"? What do you want to count in garson.config.js file?
▇ ◉ Lines
  ◉ Words
  ◯ Characters
  ◯ Everything"
`;

exports[`Multi Choice Arrow controls 8`] = `
"? What do you want to count in garson.config.js file?
▇ ◉ Lines
  ◉ Words
  ◯ Characters
  ◯ Everything"
`;

exports[`Multi Choice Initial render 1`] = `
"? What do you want to count in garson.config.js file?
▇ ◉ Lines
  ◉ Words
  ◯ Characters
  ◯ Everything"
`;

exports[`Multi Choice Toggle selection 1`] = `
"? What do you want to count in garson.config.js file?
▇ ◯ Lines
  ◉ Words
  ◯ Characters
  ◯ Everything"
`;

exports[`Multi Choice Toggle selection 2`] = `
"? What do you want to count in garson.config.js file?
  ◯ Lines
  ◉ Words
▇ ◉ Characters
  ◯ Everything"
`;

exports[`Multi Choice onChangeMiddleware 1`] = `
"? What do you want to count in garson.config.js file?
  ◯ Lines
  ◯ Words
  ◯ Characters
▇ ◉ Everything"
`;

exports[`Multi Choice onChangeMiddleware 2`] = `
"? What do you want to count in garson.config.js file?
  ◯ Lines
  ◯ Words
▇ ◉ Characters
  ◯ Everything"
`;
